domain/user/state/shared.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2024 Canonical Ltd.
```

Comment:
fwiw in the filename, having it be 'domain/user/shared/state.go' seems slightly better than 'domain/user/state/shared.go'. The latter seems to say "have a peek at my internal state, but only ever use the things that were defined in the file shared.go, which isn't very visible from a packaging perspective.

Reply:
I want to avoid that, as we'll be in the same place we're in. The state package exports public methods that the shared package can use. Why not just use the public methods directly? Conversely, the shared package will have duplicated statements, thus being able to reuse existing functions.